File: src/main/java/com/coding/exam/UnionFind.java
package com.coding.exam;

import java.util.ArrayList;
import java.util.Arrays;

public class UnionFind {
    public static void main(String[] args) {
        Solution sol = new Solution();

        Boolean[] result = sol.solution(3, new int[][]{
                {0,0,1},
                {0,1,2},
                {1,1,2}
        });
        System.out.println(Arrays.toString(result));
    }
    private static class Solution {
        private static int[] parent;

        public Boolean[] solution(int k , int[][] operations) {
            this.parent = new int[k];

            for (int i = 0; i < k; i++) {
                parent[i] = i;
            }

            ArrayList<Boolean> answer = new ArrayList<>();

            for (int[] operation : operations) {
                if (operation[0] == 0) {
                    union(operation[1], operation[2]);
                } else {
                    answer.add(find(operation[1]) == find(operation[2]));
                }
            }
            return answer.toArray(new Boolean[0]);
        }

        private static int find(int x) {
            if (parent[x] == x) { //루트노드
                return x;
            }
            return find(parent[x]);
        }
        private static void union(int x, int y) {
            int root1 = find(x);
            int root2 = find(y);
            parent[root2] = root1;
        }
    }
}

File: src/main/kotlin/com/coding/programmers/etc/UnionFind.kt
class UnionFind {
    fun solution(k: Int, operations: Array<IntArray>): Array<Boolean> {
        val parent = IntArray(k)

        //초기화
        for (i in 0 until k) {
            parent[i] = i
        }

        val ans = ArrayList<Boolean>()
        for (op in operations) {
            if (op[0] == 0) {
                union(op[1], op[2], parent)
            }
            else {
                ans.add(find(op[1], parent) == find(op[2], parent))
            }
        }
        return ans.toTypedArray()
    }

    fun union(x: Int, y: Int, parent: IntArray) {
        val root1 = find(x, parent)
        val root2 = find(y, parent)
        parent[root2] = root1
    }
    fun find(x: Int, parent: IntArray): Int {
        if (parent[x] == x) { //자기자신일 경우(루트노드 라면 x 반환
            return x
        }
        parent[x] = find(parent[x], parent)

        return parent[x]
    }
}